src/Owin.Security.OpenIdConnect.Server/OpenIdConnectServerHandler.Endpoints.cs

```diff
@@ -1729,6 +1729,81 @@ internal partial class OpenIdConnectServerHandler : AuthenticationHandler<OpenId
                 return;
             }
 
+            var usage = ticket.GetUsage();
+            var identifier = ticket.GetProperty(OpenIdConnectConstants.Extra.ClientId);
+            var audiences = ticket.GetAudiences();
+            if (string.Equals(usage, OpenIdConnectConstants.Usages.RefreshToken, StringComparison.Ordinal)) {
+                // For Refresh tokens
+                if ((clientNotification.IsSkipped &&
+                     ticket.ContainsProperty(OpenIdConnectConstants.Extra.ClientAuthenticated)) ||
+                    (!clientNotification.IsSkipped &&
+                     (!ticket.ContainsProperty(OpenIdConnectConstants.Extra.ClientAuthenticated) ||
+                      !string.Equals(identifier, clientNotification.ClientId)))) {
+                    // If the caller was not authenticated
+                    //  AND the token was issued to an authenticated client
+                    // OR if the caller was authenticated 
+                    //  AND either the token was not issued to an authenticated client OR the token was issued to a different authenticated client
+                    // Then the token belongs to a different client, reject the request.
+                    Options.Logger.WriteVerbose("refresh token issued to different client");
+
+                    await SendPayloadAsync(new JObject {
+                        [OpenIdConnectConstants.Claims.Active] = false
+                    });
+
+                    return;
+                }
+            }
+
+            else if (string.Equals(usage, OpenIdConnectConstants.Usages.IdToken, StringComparison.Ordinal)) {
+                //For Identity tokens
+                if (clientNotification.IsSkipped ||
+                    (!string.IsNullOrWhiteSpace(clientNotification.ClientId) &&
+                     !audiences.Contains(clientNotification.ClientId))) {
+                    // If the caller is not authenticated OR is not one of the specified audiences,
+                    if (ticket.ContainsProperty(OpenIdConnectConstants.Extra.ClientAuthenticated)) {
+                        // And if the token was issued to an authenticated client, it was issued to a different client
+                        Options.Logger.WriteVerbose("identity token issued to different client");
+
+                        await SendPayloadAsync(new JObject {
+                            [OpenIdConnectConstants.Claims.Active] = false
+                        });
+
+                        return;
+                    }
+                }
+            }
+
+            else if (string.Equals(usage, OpenIdConnectConstants.Usages.AccessToken, StringComparison.Ordinal)) {
+                // For access tokens
+                if (!clientNotification.IsSkipped && !string.IsNullOrWhiteSpace(clientNotification.ClientId)) {
```

Comment:
So then I'm confused. If a state is `skipped` does that mean that it's not validated, but not rejected either? That seems weird to me for whatever reason... But yeah, I just copied that part of the code from (I think) the Token endpoint and didn't change it. I'll fix that.

The simplification to `clientNotification.IsValidated' makes a bit more sense to me now though, I'll change it.